### PR TITLE
[C#] integral-numbers new exercise

### PR DESCRIPTION
**Pull request by mikedamay**

### CODE REVIEW

**Comment by mikedamay** [comment]
Please review _Example.cs_, _instructions.md_ and _introduction.md_.

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+| 65,536                     | 2,147,483,647             | int    |
+| -2,147,483,648             | -32,769                   | int    |
+| 0                          | 65,535                    | ushort |
+| -32,768                    | -1                        | short  |
+
+The value should be converted to the appropriate number of bytes for its assigned type. The complete buffer comprises a byte indicating the number of additional bytes in the buffer (_prefix byte_) followed by the bytes holding the integer (_payload bytes_).
+
+If the payload bytes represent a signed type (`short`, `int` or `long`) then the prefix byte should contain the length as a negative number.
+
+Only the prefix byte and the number of following bytes indicated by the prefix will be sent in the communication. Internally a 9 byte buffer is used (with trailing zeroes, as necessary) both by sending and receiving routines.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:21]: Reading this I got excited about doing this exercise! Love it.

File: languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+C#, like many statically typed languages, provides a number of types that represent integers, each with its own range of values. At the low end the `sbyte` type has a minimum value of -128 and a maximum value of 127. Like all the integer types these values are available as `<type>.MinValue` and `<type>.MaxValue`. At the high end the `long` type has a minimum value of -9,223,372,036,854,775,808 and a maximum value of 9,223,372,036,854,775,807. In between lie the `short` and `int` types.
+
+Each of the above types is paired with an unsigned equivalent: `sbyte`/`byte`, `short`/`ushort`, `int`/`uint` and `long`/`ulong`. In all cases the range of the values is from 0 to the negative signed maximum times 2 plus 1.
+
+A variable (or expression) of one type can easily be converted to another. For instance, in an assignment operation, if the type of the value being assigned (lhs) ensures that the value will lie within the range of the type being assigned to (rhs) then there is a simple assignment:
+
+```csharp
+ulong ul;
+uint ui = uint.MaxValue;
+ul = ui;    // no problem
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md:10]: Any particular reason not to do:

```suggestion
uint ui = uint.MaxValue;
ulong ul = ui;    // no problem
```

File: languages/csharp/exercises/concept/integral-numbers/.meta/Example.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                return BitConverter.ToInt32(buffer, 1);
+            case 2:
+                return BitConverter.ToUInt16(buffer, 1);
+            case -2:
+                return BitConverter.ToInt16(buffer, 1);
+            default:
+                return 0;
+        }
+    }
+
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.meta/Example.cs:62]:
```suggestion
```

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+## 1. Encode an integral value ready to send
+
+Please implement the static method `TelementryBuffer.ToBuffer()` to encode a buffer taking the parameter passed to the method.
+
+```csharp
+TelemetryBuffer.ToBuffer(5)
+// => {2, 5, 0, 0, 0, 0, 0, 0, 0 };
+TelemetryBuffer.ToBuffer(Int32.MaxValue)
+// => {0xfc, 0xff, 0xff, 0xff, 0x7f, 0, 0, 0, 0 };
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:31]: I was slightly confused about the mixture of hex and non-hex notation. Would it make sense to use the same notation for all bytes?

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+The value should be converted to the appropriate number of bytes for its assigned type. The complete buffer comprises a byte indicating the number of additional bytes in the buffer (_prefix byte_) followed by the bytes holding the integer (_payload bytes_).
+
+If the payload bytes represent a signed type (`short`, `int` or `long`) then the prefix byte should contain the length as a negative number.
+
+Only the prefix byte and the number of following bytes indicated by the prefix will be sent in the communication. Internally a 9 byte buffer is used (with trailing zeroes, as necessary) both by sending and receiving routines.
+
+## 1. Encode an integral value ready to send
+
+Please implement the static method `TelementryBuffer.ToBuffer()` to encode a buffer taking the parameter passed to the method.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:25]:
```suggestion
Please implement the static method `TelemetryBuffer.ToBuffer()` to encode a buffer taking the parameter passed to the method.
```

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+```csharp
+TelemetryBuffer.ToBuffer(5)
+// => {2, 5, 0, 0, 0, 0, 0, 0, 0 };
+TelemetryBuffer.ToBuffer(Int32.MaxValue)
+// => {0xfc, 0xff, 0xff, 0xff, 0x7f, 0, 0, 0, 0 };
+```
+
+## 2. Decode a received buffer
+
+Please implement the static method `TelementryBuffer.FromBuffer()` to decode the buffer received and return the value in the form of a `long`.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:36]:
```suggestion
Please implement the static method `TelemetryBuffer.FromBuffer()` to decode the buffer received and return the value in the form of a `long`.
```

File: languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
+C#, like many statically typed languages, provides a number of types that represent integers, each with its own range of values. At the low end the `sbyte` type has a minimum value of -128 and a maximum value of 127. Like all the integer types these values are available as `<type>.MinValue` and `<type>.MaxValue`. At the high end the `long` type has a minimum value of -9,223,372,036,854,775,808 and a maximum value of 9,223,372,036,854,775,807. In between lie the `short` and `int` types.
+
+Each of the above types is paired with an unsigned equivalent: `sbyte`/`byte`, `short`/`ushort`, `int`/`uint` and `long`/`ulong`. In all cases the range of the values is from 0 to the negative signed maximum times 2 plus 1.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md:3]: Would it perhaps make sense to have some sort of table listing the types and their ranges (and the number of bits/bytes?)

File: languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+```csharp
+uint ui;
+short s = 42;
+ui = (uint)s;
+```
+
+#### Bit conversion
+
+The `BitConverter` class provides a convenient way of converting integer types to and from arrays of bytes.
```
[the remark this answers is not in the document]
Comment (mikedamay) [reply languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md:23]: I think that for a C programmer knowing how to convert between an array of bits and typed values verges on a concept.  I don't think we want to waste the student's time trying to work out how to do the conversion.

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md (hunk 4 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:31]: Made consistent use of hex.  Done.

File: languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md:21]: You might be right. Let's leave it in the introduction for now

File: languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md (hunk 7 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md:3]: closed to finalize - will reopen shortly

**Comment by mikedamay** [comment]
closed to finalize - will reopen shortly

File: languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md (hunk 7 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/integral-numbers/.docs/introduction.md:3]: Done

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md (hunk 4 above)
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:31]: I don't see it in the changed files. Maybe you forgot to push the commit?

Comment (mikedamay) [reply languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:31]: You're familiar with "optimistic merging".  These are "optimistic comments".  

I will update everything and then reopen the PR.

Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:31]: > These are "optimistic comments".

I was unfamiliar with this up until now :)

**Comment by mikedamay** [comment]
All files complete and proofed.  intro includes additional spiel on casting.

File: languages/csharp/exercises/concept/integral-numbers/.docs/after.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+The multiplicity of integer types reflects machine architectures, in the size of registers, the size of CPU instruction arguments and the treatment of sign within the CPU. A value of type `long` uses 64 bits whereas a value of type `sbyte` uses 8 bits. In some cases there will be implications on CPU performance, memory usage and even disk usage (where selection of a smaller integer type will generally be beneficial). Selection of integer `type` can also be a rough and ready wsy of communicating information to other developers about the expected range of values. The `int` type is widely used as the default type where nothing special has been identified about the particular usage. The `long` or `ulong` is widely used as a simple identifier. The size of the type in bytes determines the range of values.
+
+The types discussed so far are _primitive_ types. Each is paired with a `struct` alias which implements fields (such as `MinValue`) and methods (such as `ToString()`) which are associated with the type. Some examples are: `sbyte` / `SByte`, `ushort` / `UInt16` and `long` / `Int64`.
+
+|        | Width  | Minimum                    | Maximum                     |
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/after.md:11]: Maybe also add a column for the type?

```suggestion
| Type   | Width  | Minimum                    | Maximum                     |
```

Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/integral-numbers/.docs/after.md:11]: Maybe also add the `struct` alias as a column?

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+| Min Value                  | Max Value                 | Type   |
+| -------------------------- | ------------------------- | ------ |
+| 4,294,967,296              | 9,223,372,036,854,775,807 | long   |
+| -9,223,372,036,854,775,808 | -2,147,483,649            | long   |
+| 2,147,483,648              | 4,294,967,295             | uint   |
+| 65,536                     | 2,147,483,647             | int    |
+| -2,147,483,648             | -32,769                   | int    |
+| 0                          | 65,535                    | ushort |
+| -32,768                    | -1                        | short  |
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:15]:
```suggestion
| 4,294,967,296              | 9,223,372,036,854,775,807 | `long`   |
| -9,223,372,036,854,775,808 | -2,147,483,649            | `long`   |
| 2,147,483,648              | 4,294,967,295             | `uint`   |
| 65,536                     | 2,147,483,647             | `int`    |
| -2,147,483,648             | -32,769                   | `int`    |
| 0                          | 65,535                    | `ushort` |
| -32,768                    | -1                        | `short`  |
```

File: languages/csharp/exercises/concept/integral-numbers/.docs/after.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+| ------ | ------ | -------------------------- | --------------------------- |
+| sbyte  | 8 bit  | -128                       | +127                        |
+| short  | 16 bit | -32_768                    | +32_767                     |
+| int    | 32 bit | -2_147_483_648             | +2_147_483_647              |
+| long   | 64 bit | -9_223_372_036_854_775_808 | +9_223_372_036_854_775_807  |
+| byte   | 8 bit  | 0                          | +255                        |
+| ushort | 16 bit | 0                          | +65_535                     |
+| uint   | 32 bit | 0                          | +4_294_967_295              |
+| ulong  | 64 bit | 0                          | +18_446_744_073_709_551_615 |
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/after.md:20]:
```suggestion
| `sbyte`  | 8 bit  | -128                       | +127                        |
| `short`  | 16 bit | -32_768                    | +32_767                     |
| `int`    | 32 bit | -2_147_483_648             | +2_147_483_647              |
| `long`   | 64 bit | -9_223_372_036_854_775_808 | +9_223_372_036_854_775_807  |
| `byte`   | 8 bit  | 0                          | +255                        |
| `ushort` | 16 bit | 0                          | +65_535                     |
| `uint`   | 32 bit | 0                          | +4_294_967_295              |
| `ulong`  | 64 bit | 0                          | +18_446_744_073_709_551_615 |
```

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+## 1. Encode an integral value ready to send
+
+Please implement the static method `TelemetryBuffer.ToBuffer()` to encode a buffer taking the parameter passed to the method.
+
+```csharp
+TelemetryBuffer.ToBuffer(5)
+// => {0x2, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0 };
+TelemetryBuffer.ToBuffer(Int32.MaxValue)
+// => {0xfc, 0xff, 0xff, 0xff, 0x7f, 0x0, 0x0, 0x0, 0x0 };
+```
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md:32]: Would it perhaps be useful to also have a negative number as an example?

File: languages/csharp/exercises/concept/integral-numbers/.docs/after.md (hunk 9 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/integral-numbers/.docs/after.md]: Done

File: languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md (hunk 12 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/integral-numbers/.docs/instructions.md]: Done

**Comment by mikedamay** [comment]
All review points have been addressed

**Pull request merged**